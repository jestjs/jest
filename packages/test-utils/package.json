{
  "name": "@jest/test-utils",
  "version": "30.0.0-alpha.7",
  "private": true,
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "exports": {
    ".": {
      "types": "./build/index.d.ts",
      "require": "./build/index.js",
      "import": "./build/index.mjs",
      "default": "./build/index.js"
    },
    "./package.json": "./package.json"
  },
  "dependencies": {
    "@jest/globals": "workspace:*",
    "@jest/pattern": "workspace:*",
    "@jest/types": "workspace:*",
    "@types/node": "*",
    "ansi-regex": "^5.0.1",
    "ansi-styles": "^5.0.0",
    "pretty-format": "workspace:*",
    "semver": "^7.5.3"
  },
  "devDependencies": {
    "@types/semver": "^7.1.0"
  },
  "engines": {
    "node": "^16.10.0 || ^18.12.0 || >=20.0.0"
  }
}
