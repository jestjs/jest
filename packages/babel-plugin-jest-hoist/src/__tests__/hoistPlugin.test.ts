/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as path from 'path';
import {format as formatCode, resolveConfig} from '@prettier/sync';
import pluginTester from 'babel-plugin-tester';
import {onNodeVersions} from '@jest/test-utils';
import type {Options} from 'prettier';
import babelPluginJestHoist from '..';

// We need to use the Node.js implementation of `require` to load Babel 8
// packages, instead of our sandboxed implementation, because Babel 8 is
// written in ESM and we don't support require(esm) yet.
import Module from 'node:module';
import {pathToFileURL} from 'node:url';
const createOriginalNodeRequire = Object.getPrototypeOf(Module).createRequire;
const originalNodeRequire = createOriginalNodeRequire(
  pathToFileURL(__filename),
);

const prettierOptions: Options = {
  ...resolveConfig(__filename),
  filepath: __filename,
  parser: 'babel-ts',
};

const formatResult = (code: string) => formatCode(code, prettierOptions);

describe('babel 7', () => {
  defineTests({
    babel: require('@babel/core'),
    presetReact: require('@babel/preset-react').default,
    presetTypescript: require('@babel/preset-typescript').default,
  });
});

describe('babel 8', () => {
  onNodeVersions('>=20', skipped => {
    const req = skipped ? () => null : originalNodeRequire;
    defineTests({
      babel: req('@babel-8/core'),
      presetReact: req('@babel-8/preset-react'),
      presetTypescript: req('@babel-8/preset-typescript'),
    });
  });
});

function defineTests({
  babel,
  presetReact,
  presetTypescript,
}: {
  babel: typeof import('@babel/core');
  presetReact: any;
  presetTypescript: any;
}) {
  pluginTester({
    babel,
    plugin: babelPluginJestHoist,
    pluginName: 'babel-plugin-jest-hoist',
    tests: {
      /* eslint-disable sort-keys */
      'automatic react runtime': {
        babelOptions: {
          babelrc: false,
          configFile: false,
          filename: path.resolve(__dirname, '../file.js'),
          presets: [[presetReact, {development: true, runtime: 'automatic'}]],
        },
        code: formatResult(`
          jest.mock('./App', () => () => <div>Hello world</div>);
        `),
        formatResult(code) {
          // replace the filename with something that will be the same across OSes and machine
          const codeWithoutSystemPath = code.replace(
            /var _jsxFileName = ".*";/,
            'var _jsxFileName = "/root/project/src/file.js";',
          );

          return formatResult(codeWithoutSystemPath);
        },
        snapshot: true,
      },
      'top level mocking': {
        code: formatResult(`
          require('x');

          jest.enableAutomock();
          jest.disableAutomock();
        `),
        formatResult,
        snapshot: true,
      },
      'within a block': {
        code: formatResult(`
          beforeEach(() => {
            require('x')
            jest.mock('someNode')
          })
        `),
        formatResult,
        snapshot: true,
      },
      'within a block with no siblings': {
        code: formatResult(`
          beforeEach(() => {
            jest.mock('someNode')
          })
        `),
        formatResult,
        snapshot: true,
      },

      'required `jest` within `jest`': {
        code: formatResult(`
          const {jest} = require('@jest/globals');

          jest.mock('some-module', () => {
            jest.requireActual('some-module');
          });
        `),
        formatResult,
        snapshot: true,
      },
      'imported jest.mock within jest.mock': {
        code: formatResult(`
          import {jest} from '@jest/globals';

          jest.mock('some-module', () => {
            jest.mock('some-module');
          });
        `),
        formatResult,
        snapshot: true,
      },
      'global jest.mock within jest.mock': {
        code: formatResult(`
          jest.mock('some-module', () => {
            jest.mock('some-module');
          });
        `),
        formatResult,
        snapshot: true,
      },
      'imported jest.requireActual in jest.mock': {
        code: formatResult(`
          import {jest} from '@jest/globals';

          jest.mock('some-module', () => {
            jest.requireActual('some-module');
          });

          jest.requireActual('some-module');
        `),
        formatResult,
        snapshot: true,
      },
      'global jest.requireActual in jest.mock': {
        code: formatResult(`
          jest.mock('some-module', () => {
            jest.requireActual('some-module');
          });

          jest.requireActual('some-module');
        `),
        formatResult,
        snapshot: true,
      },
      'TS typeof usage in jest.mock': {
        babelOptions: {
          babelrc: false,
          configFile: false,
          filename: path.resolve(__dirname, '../file.ts'),
          presets: [presetTypescript],
        },
        code: formatResult(`
          jest.mock('some-module', () => {
            const actual = jest.requireActual('some-module');

            return jest.fn<typeof actual.method>();
          });
        `),
        formatResult,
        snapshot: true,
      },
      'jest.spyOn call on the imported module': {
        code: formatResult(`
          jest.mock('some-module', () => {
            const module = jest.requireActual('some-module');
            jest.spyOn(module, 'add');
            return module;
          });
        `),
        formatResult,
        snapshot: true,
      },
      'jest.spyOn call in class constructor': {
        code: formatResult(`
          jest.mock('some-module', () => {
            const Actual = jest.requireActual('some-module');
            return class Mocked extends Actual {
              constructor() {
                super();
                jest.spyOn(this, 'add');
              }
            };
          });
        `),
        formatResult,
        snapshot: true,
      },
    },
    /* eslint-enable */
  });
}
