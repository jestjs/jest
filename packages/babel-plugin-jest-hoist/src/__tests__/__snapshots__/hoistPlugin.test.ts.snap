// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`babel-plugin-jest-hoist 1. automatic react runtime: 1. automatic react runtime 1`] = `

jest.mock('./App', () => () => <div>Hello world</div>);


      ↓ ↓ ↓ ↓ ↓ ↓

var _jsxFileName = '/root/project/src/file.js';
_getJestObj().mock(
  './App',
  () => () =>
    /*#__PURE__*/ _jsxDEV(
      'div',
      {
        children: 'Hello world',
      },
      void 0,
      false,
      {
        fileName: _jsxFileName,
        lineNumber: 1,
        columnNumber: 32,
      },
      this,
    ),
);
import {jsxDEV as _jsxDEV} from 'react/jsx-dev-runtime';
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}

`;

exports[`babel-plugin-jest-hoist 2. top level mocking: 2. top level mocking 1`] = `

require('x');

jest.enableAutomock();
jest.disableAutomock();


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().enableAutomock();
_getJestObj().disableAutomock();
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}
require('x');

`;

exports[`babel-plugin-jest-hoist 3. within a block: 3. within a block 1`] = `

beforeEach(() => {
  require('x');
  jest.mock('someNode');
});


      ↓ ↓ ↓ ↓ ↓ ↓

function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}
beforeEach(() => {
  _getJestObj().mock('someNode');
  require('x');
});

`;

exports[`babel-plugin-jest-hoist 4. within a block with no siblings: 4. within a block with no siblings 1`] = `

beforeEach(() => {
  jest.mock('someNode');
});


      ↓ ↓ ↓ ↓ ↓ ↓

function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}
beforeEach(() => {
  _getJestObj().mock('someNode');
});

`;

exports[`babel-plugin-jest-hoist 5. required \`jest\` within \`jest\`: 5. required \`jest\` within \`jest\` 1`] = `

const {jest} = require('@jest/globals');

jest.mock('some-module', () => {
  jest.requireActual('some-module');
});


      ↓ ↓ ↓ ↓ ↓ ↓

const {jest} = require('@jest/globals');
jest.mock('some-module', () => {
  jest.requireActual('some-module');
});

`;

exports[`babel-plugin-jest-hoist 6. imported jest.mock within jest.mock: 6. imported jest.mock within jest.mock 1`] = `

import {jest} from '@jest/globals';

jest.mock('some-module', () => {
  jest.mock('some-module');
});


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  _getJestObj().mock('some-module');
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}
import {jest} from '@jest/globals';

`;

exports[`babel-plugin-jest-hoist 7. global jest.mock within jest.mock: 7. global jest.mock within jest.mock 1`] = `

jest.mock('some-module', () => {
  jest.mock('some-module');
});


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  _getJestObj().mock('some-module');
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}

`;

exports[`babel-plugin-jest-hoist 8. imported jest.requireActual in jest.mock: 8. imported jest.requireActual in jest.mock 1`] = `

import {jest} from '@jest/globals';

jest.mock('some-module', () => {
  jest.requireActual('some-module');
});

jest.requireActual('some-module');


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  _getJestObj().requireActual('some-module');
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}
import {jest} from '@jest/globals';
jest.requireActual('some-module');

`;

exports[`babel-plugin-jest-hoist 9. global jest.requireActual in jest.mock: 9. global jest.requireActual in jest.mock 1`] = `

jest.mock('some-module', () => {
  jest.requireActual('some-module');
});

jest.requireActual('some-module');


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  _getJestObj().requireActual('some-module');
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}
jest.requireActual('some-module');

`;

exports[`babel-plugin-jest-hoist 10. TS typeof usage in jest.mock: 10. TS typeof usage in jest.mock 1`] = `

jest.mock('some-module', () => {
  const actual = jest.requireActual('some-module');

  return jest.fn<typeof actual.method>();
});


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  const actual = jest.requireActual('some-module');
  return jest.fn();
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}

`;

exports[`babel-plugin-jest-hoist 11. jest.spyOn call on the imported module: 11. jest.spyOn call on the imported module 1`] = `

jest.mock('some-module', () => {
  const module = jest.requireActual('some-module');
  jest.spyOn(module, 'add');
  return module;
});


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  const module = jest.requireActual('some-module');
  _getJestObj().spyOn(module, 'add');
  return module;
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}

`;

exports[`babel-plugin-jest-hoist 12. jest.spyOn call in class constructor: 12. jest.spyOn call in class constructor 1`] = `

jest.mock('some-module', () => {
  const Actual = jest.requireActual('some-module');
  return class Mocked extends Actual {
    constructor() {
      super();
      jest.spyOn(this, 'add');
    }
  };
});


      ↓ ↓ ↓ ↓ ↓ ↓

_getJestObj().mock('some-module', () => {
  const Actual = jest.requireActual('some-module');
  return class Mocked extends Actual {
    constructor() {
      super();
      _getJestObj().spyOn(this, 'add');
    }
  };
});
function _getJestObj() {
  const {jest} = require('@jest/globals');
  _getJestObj = () => jest;
  return jest;
}

`;
